components/resource_control/src/lib.rs

```diff
@@ -35,29 +46,78 @@ lazy_static! {
     .unwrap();
 }
 
+pub const CONFIG_PATH: &str = "resource_group";
+
 pub struct ResourceController {
     resource_groups: DashMap<String, Arc<ResourceGroup>>,
     total_cpu_quota: f64,
     last_min_vt: AtomicU64,
     start_ts: Instant,
     // the value is the duration delta(in ms) from start_ts
     last_vt_update_time: AtomicU64,
+    pd_client: Arc<dyn PdClient>,
+}
+
+pub async fn watch_resource_group(
+    mut rpc_receiver: grpcio::ClientSStreamReceiver<pdpb::WatchGlobalConfigResponse>,
+    request_tx: mpsc::Sender<ResourceGroupConfig>,
+) {
+    while let Some(Ok(r)) = rpc_receiver.next().await {
+        let change = r.get_changes();
+        for item in change {
+            match serde_json::from_str::<ResourceGroupConfig>(item.get_value()) {
+                Ok(config) => {
+                    if let Err(e) = request_tx.try_send(config) {
+                        info!("watcher worker terminated by send err"; "err is: " => ?e);
+                        return;
+                    }
+                }
+                Err(e) => {
+                    error!("failed to watch deserialize config json, err: {}", e);
+                }
+            }
+        }
+    }
 }
 
 impl ResourceController {
-    pub fn new() -> Self {
+    pub fn new(pd_client: Arc<dyn PdClient>) -> Self {
         let total_cpu_quota = SysQuota::cpu_cores_quota() * 1000.0;
         let r = Self {
             resource_groups: DashMap::new(),
             total_cpu_quota,
             last_min_vt: AtomicU64::new(0),
             start_ts: Instant::now_coarse(),
             last_vt_update_time: AtomicU64::new(0),
+            pd_client: pd_client.clone(),
         };
-        r.init_default_group();
+        let old_cfgs = r.load_global_config();
+        if old_cfgs.is_empty() {
+            r.init_default_group();
+            r.store_global_config();
+        }
+        r.start_watch();
         r
     }
 
+    fn start_watch(&self) {
```

Comment:
Can we abstract more unify way to watch the global config? maybe other model also need it not only resource control.

Reply:
get it